baybe/surrogates/utils.py

```diff
@@ -57,156 +55,101 @@ def _prepare_targets(y: Tensor) -> Tensor:
     """
-
-    class ScaledModel(*model_cls.__bases__):
-        """Overrides the methods of the given model class such the use scaled data.
-
-        It stores an instance of the underlying model class and a scalar object.
-        """
-
-        # The posterior mode is chosen to match that of the wrapped model class
-        joint_posterior: ClassVar[bool] = model_cls.joint_posterior
-        # See base class.
-
-        def __init__(self, *args, **kwargs):
-            self.model = model_cls(*args, **kwargs)
-            self.__class__.__name__ = self.model.__class__.__name__
-            self.scaler = None
-
-        def _posterior(self, candidates: Tensor) -> tuple[Tensor, Tensor]:
-            """Call the posterior function of the internal model instance.
-
-            This call is made on a scaled version of the test data and rescales the
-            output accordingly.
-            """
-            candidates = self.scaler.transform(candidates)
-            mean, covar = self.model._posterior(candidates)
-            return self.scaler.untransform(mean, covar)
-
-        def _fit(
-            self, searchspace: SearchSpace, train_x: Tensor, train_y: Tensor
-        ) -> None:
-            """Fits the scaler and the model using the scaled training data."""
-            self.scaler = DefaultScaler(searchspace.discrete.comp_rep)
-            train_x, train_y = self.scaler.fit_transform(train_x, train_y)
-            self.model.fit(searchspace, train_x, train_y)
-
-        def __getattribute__(self, attr):
-            """Access the attributes of the class instance if available.
-
-            If the attributes are not available, it uses the attributes of the internal
-            model instance.
-            """
-            # Try to retrieve the attribute in the class
-            try:
-                val = super().__getattribute__(attr)
-            except AttributeError:
-                pass
-            else:
-                return val
-
-            # If the attribute has not been overwritten, use that of the internal model
-            return self.model.__getattribute__(attr)
-
-    # Wrapping a class using a decorator does not transfer the doc, resulting in the
-    # autodocumentation not showing the correct docstring. We thus need to manually
-    # assign the docstring of the class.
-    ScaledModel.__doc__ = model_cls.__doc__
-    return ScaledModel
+    # References to original methods
+    _fit_original = cls._fit
+    _posterior_original = cls._posterior
+
+    def _posterior_new(self, candidates: Tensor) -> tuple[Tensor, Tensor]:
+        candidates = self._autoscaler.transform(candidates)
+        mean, covar = _posterior_original(self, candidates)
+        return self._autoscaler.untransform(mean, covar)
+
+    def _fit_new(
+        self, searchspace: SearchSpace, train_x: Tensor, train_y: Tensor
+    ) -> None:
+        try:
+            self._autoscaler = DefaultScaler(searchspace.discrete.comp_rep)
```

Comment:
does this not scale integer encoded cats which shouldnt be done?
I guess you intend that for future PRs dealing with the scaling `glattziehen`, so maybe a TODO or so would fit here

Reply:
I guess so, but that was also the case before the PR, right? Definitely, rigorous scaling is one of our biggest "Baustellen" in the code ... =/

Reply:
@Scienfitz: Can I close this thread? As said, you are right but I guess the problem is strictly speaking not related to the PR ...

Reply:
I suggested adding TODO, did you consider?

Reply:
no need, serialization is completely fucked up anyway :D It's perfectly clear that we need to carefully redesign the entire processing pipeline anyway, and I am quite confident to say that the entire `DefaultScaler` class will not stay in it's current form :smile: